src/p_map.c

```diff
@@ -1635,9 +1635,38 @@ static boolean PTR_ShootTraverse(intercept_t *in)
 	  // it's a sky hack wall
 	  // fix bullet-eaters -- killough:
 	  if  (li->backsector && li->backsector->ceilingpic == skyflatnum)
-	    if (demo_compatibility || li->backsector->ceilingheight < z)
-	      return false;
+    {
+      if(!mouselook && freeaim != freeaim_direct && !casual_play)
+      {
+	      if (demo_compatibility || li->backsector->ceilingheight < z)
+	        return false;
+      }
+      else
+        if (li->backsector->ceilingheight < z) // [nugget] freeaim - fix disappearing bullet puffs when outside
+          return false;
+    }
 	}
+      // [nugget] Taken from Crispy Doom - check if the bullet puff's z-coordinate is below of above
+      // its spawning sector's floor or ceiling, respectively, and move its
+      // coordinates to the point where the trajectory hits the plane
+      if (mouselook && freeaim == freeaim_direct && casual_play)
```

Comment:
Can you think of a reason _not_ to always apply this during "Casual Play"? In the other freeaim mode and even without freeaim at all, shots could still hit a ceiling. I realize it could cause desyncs on demos and net play, but outside of that I myself see no reason not to allow this bugfix. Same goes for the check just a few lines above.